### PR TITLE
Fold Qwen3.5 shared expert into the routed grouped GEMM on MUSA (+4-5% MoE decode)

### PR DESCRIPTION
**Pull request by yeahdongcn**
## Summary

On MUSA the Qwen3.5/Qwen3.6 always-on shared expert runs as a **separate serial
MLP** — the shared-expert overlap stream is unavailable (`aux_stream()` returns
`None` off CUDA), so it adds per-layer serial kernels to the decode path. This
folds the shared expert into the routed grouped GEMM as one extra always-selected
expert slot (SGLang's `num_fused_shared_experts=1` technique), so a single kernel
covers routed + shared work. Default-on for the unquantized MUSA path when the
shared and routed intermediate sizes match; numerically equivalent output.

## Why (profiling-driven)

A head-to-head torch-profiler comparison of Qwen3.6-35B-A3B bf16 decode
(SGLang leads +15% bs1 / +22% bs16, pure TPOT) found the MoE expert GEMM is
already at parity (identical Triton `fused_moe_kernel`) — the movable MoE-path
delta was the shared expert: vLLM-MUSA ran it as a separate per-layer MLP
(`KernelSwiGlu`), while SGLang folds it into `fused_moe`. Because the MUSA shared
MLP runs *serially*, its fixed device cost is not hidden by CUDAGraph replay —
so removing it is a real captured-decode win, not a launch-count mirage.

## What changed (2 files, +101/−5, default-on, no rebuild)

- `vllm/model_executor/models/qwen3_next.py` (series patch `0085`): when
  unquantized + MUSA + `shared_expert_intermediate_size == moe_intermediate_size`,
  pass `shared_experts=None` to `FusedMoE` (no separate serial MLP) and stash the
  shared MLP + gate on the routed-experts layer.
- `vllm_musa/.../unquantized_fused_moe_method.py`: `process_weights_after_loading`
  concatenates the shared `gate_up`/`down` as one extra expert slot **in place**
  (frees the old tensors — no full copy); `forward_oot` appends one routing column
  (`weight = sigmoid(shared_gate(x))`) so the shared expert rides the grouped GEMM.
  Inert unless the block stashed the shared MLP.

## E2E (drift-controlled A-B-A, Qwen3.6-35B-A3B bf16, TP4, compile ON + FULL_DECODE_ONLY, in128/out512)

Measured on the shipped GDN decode path (Option-B reroute) as the baseline:

| config | bs1 TPOT | bs1 tok/s | bs16 TPOT | bs16 tok/s |
|---|---|---|---|---|
| baseline | 15.96 ms | 62.1 | 22.23 ms | 709 |
| + shared-expert fold | **15.10 ms** | **65.6** | **21.39 ms** | **736** |
| Δ | **−5.4%** | **+5.6%** | **−3.8%** | **+3.9%** |

Cross-round noise <0.5%, A-B-A drift <0.3% — both gains are outside noise.
Capture-safe: 0 CUDAGraph violations, correct output under FULL capture.

## Correctness

Greedy (temp 0), fold vs separate-shared baseline, under capture: leading tokens
identical (France→Paris 24/24, 17+26→43 23/24, primary-colors 24/24). Divergence
only deep in the sequence (bf16 grouped-GEMM accumulation-order noise). No
NaN/poison.

## Test plan

- [x] Correctness vs the separate-shared-expert baseline (leading-token match, under capture)
- [x] E2E A-B-A on Qwen3.6-35B-A3B bf16, TP4, compile ON + FULL_DECODE_ONLY — deltas above
- [x] Capture-safety (0 violations, fold-active log confirmed)
- [x] Inert path (baseline arm = fold-off) verified correct

## Risk

Low. Gated to **MUSA + unquantized + matching intermediate sizes** — inert for
every other config (the fold only activates when the block stashes the shared
MLP). FP8 MoE uses a different (quantized) method and is structurally unaffected;
non-MoE models never reach this path. Python-only (series patch + method) — no
native rebuild. Weights folded once after load, before capture, in place.

### CODE REVIEW

File: vllm_musa/model_executor/layers/fused_moe/unquantized_fused_moe_method.py [hunk 1]
```diff
@@ -72,6 +72,10 @@ def select_gemm_impl(
                 quant_config=self.moe_quant_config,
             )
 
+    def process_weights_after_loading(self, layer: "FusedMoE") -> None:  # type: ignore[name-defined] # noqa: F821
+        super().process_weights_after_loading(layer)
+        _fold_shared_expert_weights(layer)
```
[the remark this answers is not in the document]
Comment (yeahdongcn) [reply vllm_musa/model_executor/layers/fused_moe/unquantized_fused_moe_method.py:77]: Verified correct — the block stashes `_musa_shared_mlp`/`_musa_shared_gate` on the same `FusedMoE` instance whose quant method runs `process_weights_after_loading`/`forward_oot`, so the fold that reads them and the routing extension that uses them see the same layer. Confirmed on Qwen3.6-35B-A3B (bf16, TP4): the fold activates ("MUSA shared-expert fold active … slot 256" log) with correct output. Added a clarifying comment documenting that invariant in e82bc7b38.

File: vllm_musa/model_executor/layers/fused_moe/unquantized_fused_moe_method.py [hunk 2]
```diff
@@ -81,10 +85,21 @@ def forward_oot(
...
-        # shared_experts kwargs. Return routed output here; the runner computes
-        # and combines shared experts for non-overlapped MUSA paths.
+        # The legacy fused_experts() helper computes routed experts only. When a
+        # Qwen3.5 shared expert has been folded into the expert weights as one
+        # extra always-selected slot, extend the routing by a single column so
+        # the shared expert rides the same grouped GEMM instead of running as a
+        # separate serial MLP. Otherwise the shared expert is combined by the
+        # runner outside this call.
+        if getattr(layer, "_musa_shared_folded", False):
+            shared_logits, _ = layer._musa_shared_gate(x)
```
[the remark this answers is not in the document]
Comment (yeahdongcn) [reply vllm_musa/model_executor/layers/fused_moe/unquantized_fused_moe_method.py:102]: Addressed in e82bc7b38. The fold installs `shared_experts=None` on `FusedMoE`, so the runner never supplies a separate `shared_experts_input` here — `x` is the shared expert's input, matching the routed experts it now rides with (the correctness A/B is bit-identical to the separate-shared baseline). Rather than switch the gate to a never-provided `shared_experts_input` (which would desync the gate from the x-based folded GEMM), added an `assert shared_experts_input is None` to make that invariant fail-loud instead of silently wrong.
